guppylang/std/_internal/checker.py

```diff
@@ -536,3 +538,27 @@ def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
         assert len(inst) == 0, "func_ty is not generic"
         node = BarrierExpr(args=args, func_ty=func_ty)
         return with_loc(self.node, node), ret_ty
+
+
+class StateResultChecker(CustomCallChecker):
+    """Call checker for the `state_result` function."""
+
+    def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
+        tag, _ = ExprChecker(self.ctx).check(args[0], string_type())
+        if not isinstance(tag, ast.Constant) or not isinstance(tag.value, str):
+            raise GuppyTypeError(ExpectedError(tag, "a string literal"))
+        if len(tag.value.encode("utf-8")) > TAG_MAX_LEN:
+            err: Error = TooLongError(tag)
+            err.add_sub_diagnostic(TooLongError.Hint(None))
+            raise GuppyTypeError(err)
+
+        tys = [ExprSynthesizer(self.ctx).synthesize(val)[1] for val in args]
+        func_ty = FunctionType(
+            [FuncInput(tys[0], InputFlags.NoFlags)]
+            + [FuncInput(t, InputFlags.Inout) for t in tys[1:]],
```

Comment:
Then you could just do

```suggestion
            [FuncInput(string_type(), InputFlags.NoFlags)]
            + [FuncInput(t, InputFlags.Inout) for t in tys],
```

here